File: kotlin-node/karakum/src/jsMain/resources/plugins/convertErrorTypeReferenceNode.js
import ts from "typescript";

export default function (node) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Error"
    ) {

        return "js.errors.JsError"
    }
    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "Error"
    ) {

        return "js.errors.JsError"
    }
    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/plugins/convertErrorTypeReferenceNode.kt
package electron.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isExpressionWithTypeArguments
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertErrorTypeReferenceNode = createPlugin { node, _, _ ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Error")

        "js.errors.JsError"
    } ?: nullable {
        ensure(isExpressionWithTypeArguments(node))

        val expression = node.expression
        ensure(isIdentifier(expression))
        ensure(expression.text == "Error")

        "js.errors.JsError"
    }
}
